Sensu go plugin integration

sayakju opened the pull request.

Commit:
Initial commit

Commit:
Revert testcase changes

Commit:
test case fix

Commit:
test case fix

Commit:
Removing unnecessary event check from main.go. These checks are already being performed in sensu-plugins-go-library

fguimond commented on main.go:
```diff
@@ -22,146 +17,87 @@ type HandlerConfigOption struct {
 
 type HandlerConfig struct {
+	sensu.PluginConfig
 	SlackWebhookUrl HandlerConfigOption
```
The HandlerConfigOptions can probably be replaced with just the value itself. Something like:

```
type HandlerConfig struct {
	sensu.PluginConfig
	SlackWebhookUrl string
	SlackChannel    string
	SlackUsername   string
	SlackIconUrl    string
}

fguimond commented on Gopkg.toml:
```diff
@@ -30,12 +30,8 @@
-  name = "github.com/spf13/cobra"
-  version = "0.0.3"
+  name = "github.com/sensu/sensu-plugins-go-library"
+  version = "0.2.0-alpha1"
```
As of 15 minutes ago this can be updated to `0.2.0-alpha4`

fguimond replied:
I take that back you can now use `0.2.0` since the changes have now been merged to master.

fguimond reviewed (changes requested):
Looks great overall, just a couple of small things to be addressed 😺

Commit:
Changes based on PR Feedback

nikkictl commented on main.go:
```diff
@@ -1,167 +1,97 @@
-	"log"
-	"os"
-	"path"
+	"github.com/sensu/sensu-go/types"
```
`types` is actually an alias to our new package renaming. In sensu-go, we typically import the package as `corev2 "github.com/sensu/sensu-go/api/core/v2"`. In this case, all instances of `types.Event` would be replaced with `corev2.Event`.

nikkictl replied:
There is no way you could have known this without digging into the sensu-go code. But this would be a nice improvement (for consistency sake) if we implemented this in our plugins.

nikkictl commented on main.go:
```diff
@@ -1,167 +1,97 @@
-		RunE:  run,
+func checkArgs(_ *types.Event) error {
+	if len(config.SlackWebhookUrl) == 0 {
+		return fmt.Errorf("--webhook-url or SENSU_SLACK_WEHBOOK_URL environment variable is required")
```
```suggestion
		return fmt.Errorf("--webhook-url or SENSU_SLACK_WEBHOOK_URL environment variable is required")
```

nikkictl commented on main.go:
```diff
@@ -1,167 +1,97 @@
+	slackConfigOptions = []*sensu.PluginConfigOption{
+		{
+			Path:      "webhook-url",
+			Env:       "SENSU_SLACK_WEHBOOK_URL",
```
```suggestion
			Env:       "SENSU_SLACK_WEBHOOK_URL",
```

nikkictl commented on main.go:
```diff
@@ -1,167 +1,97 @@
+
+	slackConfigOptions = []*sensu.PluginConfigOption{
+		{
+			Path:      "webhook-url",
```
Since the `Path` and `Argument` fields for all of these options are the same, I would consider defining constants for them.

nikkictl commented on main.go:
```diff
@@ -1,167 +1,97 @@
+	goHandler, _ := sensu.NewGoHandler(&config.PluginConfig, slackConfigOptions, checkArgs, executeHandler)
+	err := goHandler.Execute()
+	if err != nil {
+		fmt.Printf("Error executing plugin: %s", err)
```
Could you explain the reasoning for the change from `log.Fatal` to `fmt.Printf`?

sayakju replied:
After upgrading to use "0.2.0" version of "sensu-plugins-go-library" library, the printing of error is now being handled by the library itself. It simply logs the error to stderr and then exits with status 1.

nikkictl commented on main.go:
```diff
@@ -1,167 +1,97 @@
-
-	if err = sendMessage(event); err != nil {
+func executeHandler(event *types.Event) error {
+	if err := sendMessage(event); err != nil {
```
This entire function could probably be simplified to `return sendMessage(event)`.. which makes me wonder why the `executeHandler` function exists anymore.

nikkictl commented:
@sayakju would you mind providing a brief description around the context of these changes? This information would be useful in the PR description, as well as in the changelog.

Commit:
Changes based on review comments:

1) Upgrade of sensu-plugins-go-library to version 0.2.0
2) Use of "github.com/sensu/sensu-go/api/core/v2" library instead of "github.com/sensu/sensu-go/types" to maintain consistency with sensu-go code

Commit:
Changes based on review comments:

3) Rectification of spelling error

sayakju commented:
Update `sensu-slack-handler` to use new `sensu-plugins-go-library`

nikkictl commented on Gopkg.toml:
```diff
@@ -30,12 +30,8 @@
   name = "github.com/bluele/slack"
 
 [[constraint]]
-  name = "github.com/sensu/sensu-go"
```
It appears that we're still importing our corev2 package from sensu-go. Is there a specific reason we're removing it from our vendor directory/dependency management?

sayakju replied:
sensu/sensu-pagerduty-handler#14 (comment)

nikkictl commented on main_test.go:
```diff
@@ -129,14 +129,14 @@ func TestMain(t *testing.T) {
 	_, err = file.Seek(0, 0)
 	require.NoError(t, err)
-	stdin = file
+	os.Stdin = file
```
It's safer for test stability to avoid reading/writing files unless absolutely necessary (see https://github.com/sensu/sensu-plugins-go-library/pull/5#discussion_r286107307). Would it be possible to stream these bytes to stdin instead?

sayakju replied:
While you certainly make sense, but I am finding it difficult to stream bytes to os.Stdin, while is of type *File. Since, this is something which was pre-existing and not part of the change done in this PR, can it be taken care of by some other pull request?

nikkictl reviewed (commented):
Looks good! I have a few questions 👍

Commit:
Adding direct dependency on github.com/sensu/sensu-go library

nikkictl reviewed (approved):
Thanks for adding the direct dependency! If you'd like to make the testing changes in another PR thats OK 😄

fguimond commented:
@nikkiki @calebhailey - @sayakju and I don't have permissions to merge the changes. Can you please merge them when you have a chance so we can resolve that item? Or please let us know what the process should be.

nikkictl commented:
@fguimond sure, I can go ahead and merge this. I've made an action item for the team to discuss getting y'all the proper permissions, and to organize our review process a little better for these repos.

fguimond commented:
Awesome, this will definitely help!
